### PR TITLE
Add support for "show grant"

### PR DESCRIPTION
**Pull request by amrutagokhale**
This PR adds syntax support for "show grant" and its implementation for Hive connector. This command is not specified in SQL standard, so there there is no gold standard for the syntax. I have borrowed the syntax from Hive. Here is the syntax that the current version supports:
`SHOW GRANT
        (USER identifier)?
        ON (TABLE? qualifiedName | ALL)`

It supports showing grants for users but lacks the support to show grants for roles (which I am planning to add later).
- When user name is speciified, `SHOW GRANT USER identifier ...` will show grants for the specified user, provided the specified user is the same as the current user. Attempts to show grants for any other user than the current user will be denied.
- When user is not specified, `SHOW GRANT ...` will show grants for the current user. 

The implementation follows the definition of `TABLE_PRIVILEGES` table specified in SQL standard:

``` sql
CREATE VIEW TABLE_PRIVILEGES AS
SELECT GRANTOR, GRANTEE, TABLE_CATALOG, TABLE_SCHEMA, TABLE_NAME,
PRIVILEGE_TYPE, IS_GRANTABLE, WITH_HIERARCHY
FROM DEFINITION_SCHEMA.TABLE_PRIVILEGES
WHERE ( GRANTEE IN
( ’PUBLIC’, CURRENT_USER )
OR
GRANTOR = CURRENT_USER )
AND
TABLE_CATALOG
= ( SELECT CATALOG_NAME
FROM INFORMATION_SCHEMA_CATALOG_NAME );
```

except that:
1) We don't consider the case of grantor = current_user
2) grantee in `PUBLIC` is not taken into consideration

I will handle these cases soon.

PS: I will add product tests and documentation in a separate PR once this gets in.

### CODE REVIEW

File: presto-parser/src/main/antlr4/com/facebook/presto/sql/parser/SqlBase.g4 [hunk 1]
```diff
@@ -55,6 +55,9 @@ statement
         (GRANT OPTION FOR)?
         (privilege (',' privilege)* | ALL PRIVILEGES)
         ON TABLE? qualifiedName FROM grantee=identifier                #revoke
+    | SHOW GRANT
+        (USER identifier)?
```
Comment (cawallin) [inline presto-parser/src/main/antlr4/com/facebook/presto/sql/parser/SqlBase.g4]: I'd suggest removing `(USER identifier)?` for now, since if I understand your description, it'll just fail if the identifier isn't current user.

File: presto-parser/src/main/java/com/facebook/presto/sql/SqlFormatter.java [hunk 2]
```diff
@@ -978,6 +979,35 @@ public Void visitRevoke(Revoke node, Integer indent)
...
+        {
+            builder.append("SHOW GRANT ");
+            if (node.getIdentityType().isPresent()) {
+                if (node.getIdentityType().get() == ShowGrant.IdentityType.USER) {
+                    builder.append("USER ")
+                            .append(node.getIdentity().get())
+                            .append(" ");
+                }
+                else {
+                    throw new UnsupportedOperationException("unhandled identity : " + node.getIdentityType());
```
Comment (cawallin) [inline presto-parser/src/main/java/com/facebook/presto/sql/SqlFormatter.java:993]: Use String.format()
Also, mention something about it being in SHOW GRANT.
Suggest: "Unsupported identity type in SHOW GRANTS (only USER supported): %s"

File: presto-parser/src/main/java/com/facebook/presto/sql/parser/AstBuilder.java [hunk 3]
```diff
@@ -701,6 +702,31 @@ public Node visitRevoke(SqlBaseParser.RevokeContext context)
...
+        Optional<String> identity = Optional.empty();
+        Optional<QualifiedName> tableName = Optional.empty();
+
+        if (context.USER() != null) {
+            identityType = Optional.of(ShowGrant.IdentityType.USER);
+            identity = Optional.of(context.identifier().getText());
+        }
+
+        if (context.ALL() == null) {
+            tableName = Optional.of(getQualifiedName(context.qualifiedName()));
```
Comment (cawallin) [inline presto-parser/src/main/java/com/facebook/presto/sql/parser/AstBuilder.java:718]: the `TABLE` keyword is optional, does that mean that this isn't necessarily a tablename? could it be a view or a schema?

File: presto-parser/src/main/java/com/facebook/presto/sql/tree/ShowGrant.java [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+    private final boolean table;
+    private final Optional<QualifiedName> tableName;
+    private final boolean all;
+
+    public ShowGrant(Optional<IdentityType> identityType, Optional<String> identity, boolean table, Optional<QualifiedName> tableName, boolean all)
+    {
+        this(Optional.empty(), identityType, identity, table, tableName, all);
+    }
+
+        public ShowGrant(NodeLocation location, Optional<IdentityType> identityType, Optional<String> identity, boolean table, Optional<QualifiedName> tableName, boolean all)
```
Comment (cawallin) [inline presto-parser/src/main/java/com/facebook/presto/sql/tree/ShowGrant.java]: The spacing looks off here, make sure to use the Airlift intellij codestyle plugin

File: presto-main/src/main/java/com/facebook/presto/connector/informationSchema/InformationSchemaMetadata.java [hunk 5]
```diff
@@ -95,6 +97,14 @@
                     .column("partition_key", createUnboundedVarcharType())
                     .column("partition_value", createUnboundedVarcharType())
                     .build())
+            .table(tableMetadataBuilder(TABLE_TABLE_PRIVILEGES)
```
Comment (cawallin) [inline presto-main/src/main/java/com/facebook/presto/connector/informationSchema/InformationSchemaMetadata.java:100]: what about "grantor" and "with_hierarchy"? do we not have access to that information?

File: presto-main/src/main/java/com/facebook/presto/connector/informationSchema/InformationSchemaPageSourceProvider.java [hunk 6]
```diff
@@ -320,6 +352,13 @@ private static QualifiedTablePrefix extractQualifiedTablePrefix(String catalogNa
         return new QualifiedTablePrefix(catalogName, schemaName, tableName);
     }
 
+    private static String extractGrantee(Session session, String catalogName, Map<String, NullableValue> filters)
+    {
+        Optional<String> grantee = getFilterColumn(filters, "grantee");
+        checkArgument(grantee.isPresent(), "filter is required for column: %s.%s", TABLE_TABLE_PRIVILEGES, "grantee");
```
Comment (cawallin) [inline presto-main/src/main/java/com/facebook/presto/connector/informationSchema/InformationSchemaPageSourceProvider.java]: I'm not sure you can enforce this filter, these information schema tables are sometimes used by JDBC/ODBC drivers and/or BI tools. You could default to the current user, though.

File: presto-main/src/main/java/com/facebook/presto/connector/informationSchema/InformationSchemaPageSourceProvider.java [hunk 7]
```diff
@@ -204,6 +210,32 @@ private InternalTable buildTables(Session session, String catalogName, Map<Strin
...
+                            privilegeInfo.isGrantOption());
+                }
+            }
+        }
+        return table.build();
+    }
+
+    private List<GrantInfo> getTablePrivilegesList(Session session, String catalogName, Map<String, NullableValue> filters)
+    {
+        return metadata.listTablePrivileges(session, extractQualifiedTablePrefix(catalogName, filters), extractGrantee(session, catalogName, filters));
```
Comment (cawallin) [inline presto-main/src/main/java/com/facebook/presto/connector/informationSchema/InformationSchemaPageSourceProvider.java]: beware, the query won't necessarily be on only one table, it could span several

Comment (cawallin) [reply presto-main/src/main/java/com/facebook/presto/connector/informationSchema/InformationSchemaPageSourceProvider.java]: Also, how will this behave if a catalog doesn't support grant/revoke? Make sure it doesn't throw an exception, maybe it should give some realistic information based on the tables in the query (aka open privileges?). What do you think?

File: presto-main/src/main/java/com/facebook/presto/sql/rewrite/ShowQueriesRewrite.java [hunk 8]
```diff
@@ -201,6 +204,42 @@ protected Node visitShowTables(ShowTables showTables, Void context)
...
+
+            Optional<QualifiedName> tableName = showGrant.getTableName();
+            if (tableName.isPresent()) {
+                QualifiedObjectName qualifiedTableName = createQualifiedObjectName(session, showGrant, tableName.get());
+
+                Expression tablePredicate = equal(nameReference("table_name"), new StringLiteral(qualifiedTableName.getObjectName()));
+                predicate = logicalAnd(predicate, tablePredicate);
+            }
+
+            if (catalogName == null) {
```
Comment (cawallin) [inline presto-main/src/main/java/com/facebook/presto/sql/rewrite/ShowQueriesRewrite.java:227]: If you're doing SHOW GRANT ON ALL, do you need the catalog set?
You should also check if the catalog is included in QualifiedObjectName, if it's on a table/object.

File: presto-main/src/main/java/com/facebook/presto/sql/rewrite/ShowQueriesRewrite.java [hunk 9]
```diff
@@ -201,6 +204,42 @@ protected Node visitShowTables(ShowTables showTables, Void context)
         }
 
         @Override
+        protected Node visitShowGrant(ShowGrant showGrant, Void context)
```
Comment (cawallin) [inline presto-main/src/main/java/com/facebook/presto/sql/rewrite/ShowQueriesRewrite.java]: Pull this change out into a separate commit, please. And move the rest of this commit after the Hive change so that it compiles.

File: presto-hive/src/main/java/com/facebook/presto/hive/HiveMetadata.java [hunk 10]
```diff
@@ -1532,6 +1536,32 @@ public void revokeTablePrivileges(ConnectorSession session, SchemaTableName sche
     }
 
     @Override
+    public List<GrantInfo> listTablePrivileges(ConnectorSession session, String schemaNameOrNull, String grantee)
```
Comment (cawallin) [inline presto-hive/src/main/java/com/facebook/presto/hive/HiveMetadata.java]: The method in Metadata is `listTablePrivileges(Session session, QualifiedTablePrefix prefix, String grantee)`

File: presto-hive/src/main/java/com/facebook/presto/hive/HiveMetadata.java [hunk 11]
```diff
@@ -1532,6 +1536,32 @@ public void revokeTablePrivileges(ConnectorSession session, SchemaTableName sche
...
+    {
+        ImmutableList.Builder<GrantInfo> grantMetadata = ImmutableList.builder();
+        for (String schemaName : listSchemas(session, schemaNameOrNull)) {
```
Comment (cawallin) [inline presto-hive/src/main/java/com/facebook/presto/hive/HiveMetadata.java]: You can use the metastore method `listTables(session, qualifiedTablePrefix)` for iterating over both schemas and tables. Also, how are views handled?

File: presto-hive/src/main/java/com/facebook/presto/hive/HiveMetadata.java [hunk 12]
```diff
@@ -1532,6 +1536,32 @@ public void revokeTablePrivileges(ConnectorSession session, SchemaTableName sche
...
+            Optional<List<String>> tables = metastore.getAllTables(schemaName);
+            for (String table : tables.orElse(emptyList())) {
+                Set<HivePrivilegeInfo> hivePrivileges = metastore.getTablePrivileges(grantee, schemaName, table);
+
+                if (!hivePrivileges.isEmpty()) {
+                    ImmutableSet.Builder<PrivilegeInfo> privilegeInfoBuilder = ImmutableSet.builder();
```
Comment (cawallin) [inline presto-hive/src/main/java/com/facebook/presto/hive/HiveMetadata.java]: You can express this more concisely with the streaming API

File: presto-spi/src/main/java/com/facebook/presto/spi/security/GrantInfo.java [hunk 13]
```diff
@@ -0,0 +1,69 @@
...
+import java.util.Objects;
+import java.util.Set;
+
+public class GrantInfo
+{
+    private final Set<PrivilegeInfo> privilegeInfo;
+    private final Identity identity;
+    private final SchemaTableName schemaTableName;
+
+    public GrantInfo(Set<PrivilegeInfo> privilegeInfo, Identity identity, SchemaTableName schemaTableName)
```
Comment (cawallin) [inline presto-spi/src/main/java/com/facebook/presto/spi/security/GrantInfo.java]: You should store the catalog as well

File: presto-main/src/main/java/com/facebook/presto/metadata/MetadataManager.java [hunk 14]
```diff
@@ -713,6 +715,22 @@ public void revokeTablePrivileges(Session session, QualifiedObjectName tableName
...
+    {
+        requireNonNull(prefix, "prefix is null");
+
+        String schemaNameOrNull = prefix.getSchemaName().orElse(null);
+        Set<GrantInfo> grants = new LinkedHashSet<>();
+        for (ConnectorEntry entry : allConnectorsFor(prefix.getCatalogName())) {
+            ConnectorMetadata metadata = entry.getMetadata(session);
+            ConnectorSession connectorSession = session.toConnectorSession(entry.getCatalog());
+            grants.addAll(metadata.listTablePrivileges(connectorSession, schemaNameOrNull, grantee).stream()
+                    .collect(toList()));
```
Comment (cawallin) [inline presto-main/src/main/java/com/facebook/presto/metadata/MetadataManager.java]: toSet()?

File: presto-main/src/main/java/com/facebook/presto/metadata/Metadata.java [hunk 15]
```diff
@@ -246,6 +247,11 @@
      */
     void revokeTablePrivileges(Session session, QualifiedObjectName tableName, Set<Privilege> privileges, String grantee, boolean grantOption);
 
+    /**
+     * Gets the privileges for the specified table available to the given grantee
+     */
+    List<GrantInfo> listTablePrivileges(Session session, QualifiedTablePrefix prefix, String grantee);
```
Comment (cawallin) [inline presto-main/src/main/java/com/facebook/presto/metadata/Metadata.java:253]: Pull the SPI changes for Metadata and ConnectorMetadata into a separate commit, please.

**Review by cawallin — CHANGES_REQUESTED** [review]

File: presto-parser/src/main/antlr4/com/facebook/presto/sql/parser/SqlBase.g4 (hunk 1 above)
Comment (amrutagokhale) [reply presto-parser/src/main/antlr4/com/facebook/presto/sql/parser/SqlBase.g4]: I could do that. I kept it to distinguish between this case and the case of `ROLE identifier` which I plan to add later.

File: presto-parser/src/main/java/com/facebook/presto/sql/SqlFormatter.java (hunk 2 above)
Comment (amrutagokhale) [reply presto-parser/src/main/java/com/facebook/presto/sql/SqlFormatter.java:993]: I think this exception is to catch implementation errors, and not end user syntax errors. So, the information that this exception is coming from `visitShowGrant()` method should be sufficient. (I followed other methods in this class that throw exceptions such as `visitJoin` etc.)

File: presto-parser/src/main/java/com/facebook/presto/sql/parser/AstBuilder.java (hunk 3 above)
Comment (amrutagokhale) [reply presto-parser/src/main/java/com/facebook/presto/sql/parser/AstBuilder.java:718]: The current implementation handles only table names and doesn't handle schema names. The current `grant/revoke` methods also don't handle schemas (and views). So we could handle schemas later?

File: presto-parser/src/main/java/com/facebook/presto/sql/tree/ShowGrant.java (hunk 4 above)
Comment (amrutagokhale) [reply presto-parser/src/main/java/com/facebook/presto/sql/tree/ShowGrant.java]: corrected

File: presto-main/src/main/java/com/facebook/presto/connector/informationSchema/InformationSchemaPageSourceProvider.java (hunk 6 above)
Comment (amrutagokhale) [reply presto-main/src/main/java/com/facebook/presto/connector/informationSchema/InformationSchemaPageSourceProvider.java]: Good catch. I forgot to remove this. Modified to return the current user.

File: presto-main/src/main/java/com/facebook/presto/sql/rewrite/ShowQueriesRewrite.java (hunk 8 above)
Comment (amrutagokhale) [reply presto-main/src/main/java/com/facebook/presto/sql/rewrite/ShowQueriesRewrite.java:227]: Good point. I extracted the catalog name from QualifiedObjectName when applicable. Yes, if you're doing SHOW GRANT ON ALL, we need the catalog set.

File: presto-spi/src/main/java/com/facebook/presto/spi/security/GrantInfo.java (hunk 13 above)
Comment (amrutagokhale) [reply presto-spi/src/main/java/com/facebook/presto/spi/security/GrantInfo.java]: I could, but it won't be used anywhere. When building the table in `INFORMATION_SCHEMA`, we have the catalog name information, so I am populating from there. Thoughts?

File: presto-hive/src/main/java/com/facebook/presto/hive/HiveMetadata.java (hunk 10 above)
Comment (amrutagokhale) [reply presto-hive/src/main/java/com/facebook/presto/hive/HiveMetadata.java]: Well, this method overrides method in `ConnectorMetadata.java` and not the one in `Metadata.java`.

File: presto-main/src/main/java/com/facebook/presto/metadata/MetadataManager.java (hunk 14 above)
Comment (amrutagokhale) [reply presto-main/src/main/java/com/facebook/presto/metadata/MetadataManager.java]: done

File: presto-parser/src/main/java/com/facebook/presto/sql/parser/AstBuilder.java (hunk 3 above)
Comment (cawallin) [reply presto-parser/src/main/java/com/facebook/presto/sql/parser/AstBuilder.java:718]: ok, sounds like a plan.

File: presto-hive/src/main/java/com/facebook/presto/hive/HiveMetadata.java (hunk 10 above)
Comment (cawallin) [reply presto-hive/src/main/java/com/facebook/presto/hive/HiveMetadata.java]: Oh, I think my internet was flaky when leaving this comment: I think it would be better to use QualifiedTablePrefix here, to match with other information schema methods (e.g. the list\* methods)

File: presto-spi/src/main/java/com/facebook/presto/spi/security/GrantInfo.java (hunk 13 above)
Comment (cawallin) [reply presto-spi/src/main/java/com/facebook/presto/spi/security/GrantInfo.java]: hmm interesting. Theoretically you should be able to ask for all of the grants a user has across all of the catalogs -- SHOW GRANT ON ALL seems like it should return that.

File: presto-hive/src/main/java/com/facebook/presto/hive/HiveMetadata.java (hunk 10 above)
Comment (amrutagokhale) [reply presto-hive/src/main/java/com/facebook/presto/hive/HiveMetadata.java]: If you see in `ConnectorMetadata.java`, the other list\* methods take `String` rather than `QualifiedTablePrefix`. The methods in `Metadata.java` use `QualifiedTablePrefix` (also, this method returns privileges for all tables in the schema, rather than privileges on a single table).

File: presto-hive/src/main/java/com/facebook/presto/hive/HiveMetadata.java (hunk 11 above)
Comment (cawallin) [reply presto-hive/src/main/java/com/facebook/presto/hive/HiveMetadata.java]: Nvm, I think I was looking at the Metadata methods.

File: presto-hive/src/main/java/com/facebook/presto/hive/HiveMetadata.java (hunk 10 above)
Comment (cawallin) [reply presto-hive/src/main/java/com/facebook/presto/hive/HiveMetadata.java]: Ah, my bad. But listTableColumns takes a `SchemaTablePrefix`, I think that's what I'm remembering -- you probably want to use that so that, if necessary, you can do the requisite filtering to return privileges from one or more tables.

File: presto-spi/src/main/java/com/facebook/presto/spi/security/GrantInfo.java (hunk 13 above)
Comment (amrutagokhale) [reply presto-spi/src/main/java/com/facebook/presto/spi/security/GrantInfo.java]: Since `INFORMATION_SCHEMA` schema is catalog specific, the table `TABLE_PRIVILEGES` in that schema can only store privileges for that catalog.

Comment (cawallin) [reply presto-spi/src/main/java/com/facebook/presto/spi/security/GrantInfo.java]: ah, that bit seems reasonable. So GrantInfo wouldn't need to store the catalog. If you could get SHOW GRANT ON ALL to show all of the catalogs' grants, that'd be great, but since only one catalog actually implements grants it's a bit irrelevant right now.

File: presto-main/src/main/java/com/facebook/presto/connector/informationSchema/InformationSchemaMetadata.java (hunk 5 above)
Comment (amrutagokhale) [reply presto-main/src/main/java/com/facebook/presto/connector/informationSchema/InformationSchemaMetadata.java:100]: yeah, I can't find that information in the `PrivilegeGrantInfo` object in the Hive metastore API.

File: presto-spi/src/main/java/com/facebook/presto/spi/security/GrantInfo.java (hunk 13 above)
Comment (amrutagokhale) [reply presto-spi/src/main/java/com/facebook/presto/spi/security/GrantInfo.java]: Yeah, that's right. It's relevant only for Hive connect at the moment.

File: presto-hive/src/main/java/com/facebook/presto/hive/HiveMetadata.java (hunk 10 above)
Comment (amrutagokhale) [reply presto-hive/src/main/java/com/facebook/presto/hive/HiveMetadata.java]: yeah, that's true about `listTableColumns`. What if I rename this method to `listAllTablePrivileges`, since that's what the method is doing actually?

Comment (amrutagokhale) [reply presto-hive/src/main/java/com/facebook/presto/hive/HiveMetadata.java]: As Christina pointed out to me offline, there is a good use case for changing the function signature (such as when directly querying the `information_schema`). So changing this to take `SchemaTablePrefix`.

File: presto-main/src/main/java/com/facebook/presto/connector/informationSchema/InformationSchemaPageSourceProvider.java (hunk 7 above)
Comment (amrutagokhale) [reply presto-main/src/main/java/com/facebook/presto/connector/informationSchema/InformationSchemaPageSourceProvider.java]: As of now, it will throw an exception for all catalogs except Hive. Not sure what information to show. 
Regrading first comment: yeah, that should be taken care of.

**Comment by amrutagokhale** [comment]
@cawallin I have addressed your comments (except the part of removing `USER identifier` from the syntax).

File: presto-main/src/main/java/com/facebook/presto/connector/informationSchema/InformationSchemaPageSourceProvider.java (hunk 7 above)
Comment (cawallin) [reply presto-main/src/main/java/com/facebook/presto/connector/informationSchema/InformationSchemaPageSourceProvider.java]: I don't think that it should throw an exception (e.g. for BI tool usage, that would be bad). Maybe it could just not show any grants?

Comment (amrutagokhale) [reply presto-main/src/main/java/com/facebook/presto/connector/informationSchema/InformationSchemaPageSourceProvider.java]: Ok. Changed it to return an empty list.

File: presto-main/src/main/java/com/facebook/presto/server/CoordinatorModule.java [hunk 16]
```diff
@@ -219,6 +220,7 @@ protected void setup(Binder binder)
         executionBinder.addBinding(ShowCatalogs.class).to(SqlQueryExecutionFactory.class).in(Scopes.SINGLETON);
         executionBinder.addBinding(Use.class).to(SqlQueryExecutionFactory.class).in(Scopes.SINGLETON);
         executionBinder.addBinding(ShowSession.class).to(SqlQueryExecutionFactory.class).in(Scopes.SINGLETON);
+        executionBinder.addBinding(ShowGrant.class).to(SqlQueryExecutionFactory.class).in(Scopes.SINGLETON);
```
Comment (cawallin) [inline presto-main/src/main/java/com/facebook/presto/server/CoordinatorModule.java]: This belongs in a previous commit

File: presto-spi/src/main/java/com/facebook/presto/spi/ConnectorMetadata.java [hunk 17]
```diff
@@ -301,4 +302,12 @@ default void revokeTablePrivileges(ConnectorSession session, SchemaTableName tab
     {
         throw new PrestoException(NOT_SUPPORTED, "This connector does not support revokes");
     }
+
+    /**
+     * List table privileges granted to the specified grantee for the tables that have the specified prefix
+     */
+    default List<GrantInfo> listTablePrivileges(ConnectorSession session, SchemaTablePrefix prefix, String grantee)
```
Comment (cawallin) [inline presto-spi/src/main/java/com/facebook/presto/spi/ConnectorMetadata.java]: Don't modify this class, it's deprecated (and I think there's a PR to remove it)

File: presto-spi/src/main/java/com/facebook/presto/spi/classloader/ClassLoaderSafeConnectorMetadata.java [hunk 18]
```diff
@@ -305,6 +306,14 @@ public void revokeTablePrivileges(ConnectorSession session, SchemaTableName tabl
     }
 
     @Override
+    public List<GrantInfo> listTablePrivileges(ConnectorSession session, SchemaTablePrefix prefix, String grantee)
```
Comment (cawallin) [inline presto-spi/src/main/java/com/facebook/presto/spi/classloader/ClassLoaderSafeConnectorMetadata.java]: I think this is also deprecated

File: presto-main/src/main/java/com/facebook/presto/sql/rewrite/ShowQueriesRewrite.java (hunk 8 above)
Comment (cawallin) [reply presto-main/src/main/java/com/facebook/presto/sql/rewrite/ShowQueriesRewrite.java:227]: Is there any way to iterate over the different catalog's information schemas in the query? That would be clearer for the expected behavior of SHOW GRANT ON ALL. Otherwise, it needs to be clearly documented.

File: presto-main/src/main/java/com/facebook/presto/connector/informationSchema/InformationSchemaMetadata.java [hunk 19]
```diff
@@ -58,6 +59,7 @@
     public static final SchemaTableName TABLE_VIEWS = new SchemaTableName(INFORMATION_SCHEMA, "views");
     public static final SchemaTableName TABLE_SCHEMATA = new SchemaTableName(INFORMATION_SCHEMA, "schemata");
     public static final SchemaTableName TABLE_INTERNAL_PARTITIONS = new SchemaTableName(INFORMATION_SCHEMA, "__internal_partitions__");
+    public static final SchemaTableName TABLE_TABLE_PRIVILEGES = new SchemaTableName(INFORMATION_SCHEMA, "table_privileges");
```
Comment (cawallin) [inline presto-main/src/main/java/com/facebook/presto/connector/informationSchema/InformationSchemaMetadata.java:62]: This commit needs to go before the rewrite

File: presto-main/src/main/java/com/facebook/presto/connector/informationSchema/InformationSchemaMetadata.java (hunk 5 above)
Comment (cawallin) [reply presto-main/src/main/java/com/facebook/presto/connector/informationSchema/InformationSchemaMetadata.java:100]: I'd say you should keep them in information_schema but with a null value.

**Review by cawallin — CHANGES_REQUESTED** [review]
A few more comments.

Also, please add some sort of tests with this commit -- the product tests infra for this change should already be there b/c of the GRANT/REVOKE tests, so a smoketest there shouldn't be too hard and wouldn't change very much if there were further changes requested.

File: presto-spi/src/main/java/com/facebook/presto/spi/ConnectorMetadata.java (hunk 17 above)
Comment (amrutagokhale) [reply presto-spi/src/main/java/com/facebook/presto/spi/ConnectorMetadata.java]: I guess until it's actually removed, I have to add this. Got compilation error otherwise because of the invocation of this method in `transaction/LegacyConnectorMetadata.java`.

Comment (cawallin) [reply presto-spi/src/main/java/com/facebook/presto/spi/ConnectorMetadata.java]: You can return an empty list in LegacyConnectorMetadata

File: presto-main/src/main/java/com/facebook/presto/connector/informationSchema/InformationSchemaMetadata.java (hunk 19 above)
Comment (amrutagokhale) [reply presto-main/src/main/java/com/facebook/presto/connector/informationSchema/InformationSchemaMetadata.java:62]: The order of branches seems to be messed up here, in spite of having them in the right order locally.

Comment (cawallin) [reply presto-main/src/main/java/com/facebook/presto/connector/informationSchema/InformationSchemaMetadata.java:62]: Oh ok, that's fine, sometimes the github UI has a bug.

File: presto-main/src/main/java/com/facebook/presto/sql/rewrite/ShowQueriesRewrite.java (hunk 8 above)
Comment (amrutagokhale) [reply presto-main/src/main/java/com/facebook/presto/sql/rewrite/ShowQueriesRewrite.java:227]: I don't know of such a way, but will check again. Will keep this in mind when documenting.

**Comment by amrutagokhale** [comment]
@cawallin Addressed your comments. Also added product test for SHOW GRANT.

**Comment by amrutagokhale** [comment]
I am going to close this PR and will open a new one against sprint-43.
